zeebe/atomix/cluster/src/main/java/io/atomix/cluster/messaging/impl/HeartbeatHandler.java

```diff
@@ -0,0 +1,130 @@
+
+    Server(final Logger log) {
+      super(log);
+    }
+
+    @Override
+    public void channelRead(final ChannelHandlerContext ctx, final Object msg) {
+      if (msg instanceof final ProtocolRequest request
+          && request.subject().equals(HEARTBEAT_SUBJECT)) {
+        receivedHeartbeat = true;
+        ctx.writeAndFlush(createHeartbeat(request.id()));
+        if (!Arrays.equals(request.payload(), HEARTBEAT_PAYLOAD)) {
+          log.warn(
+              "Received unexpected heartbeat payload from {}, perhaps the message subject {} is accidentally reused",
+              request.sender(),
+              HEARTBEAT_SUBJECT);
+        }
+      }
+
+      // Pass the message to the next handler, even though it's a heartbeat.
+      ctx.fireChannelRead(msg);
+    }
+
+    @Override
+    public void userEventTriggered(final ChannelHandlerContext ctx, final Object evt) {
+      if (!(evt instanceof final IdleStateEvent idleStateEvent)) {
+        return;
+      }
+      if (idleStateEvent.state() == IdleState.READER_IDLE) {
+        // don't close the connection if we did not receive any heartbeat
+        if (receivedHeartbeat) {
+          log.warn("Connection {} timed out on the server, closing channel", ctx.channel());
+          ctx.close();
+        }
+      }
+    }
+
+    private ProtocolMessage createHeartbeat(final long id) {
+      return new ProtocolReply(id, HEARTBEAT_PAYLOAD, Status.OK);
+    }
+  }
+
+  static final class Client extends HeartbeatHandler {
+    private final LongHashSet outstandingHeartbeats = new LongHashSet();
+    private final AtomicLong messageIdGenerator;
+    private final Address advertisedAddress;
+
+    Client(final Logger log, final AtomicLong messageIdGenerator, final Address advertisedAddress) {
+      super(log);
+      this.messageIdGenerator = messageIdGenerator;
+      this.advertisedAddress = advertisedAddress;
+    }
+
+    @Override
+    public void channelRead(final ChannelHandlerContext ctx, final Object msg) {
+      if (msg instanceof final ProtocolReply reply && outstandingHeartbeats.contains(reply.id())) {
+        // remove all heartbeats sent before the last update
+        outstandingHeartbeats.removeIfLong(id -> id <= reply.id());
+        if (reply.status() != Status.OK) {
+          log.warn("Received a Heartbeat response with status {}", reply.status());
+        } else if (!Arrays.equals(reply.payload(), HEARTBEAT_PAYLOAD)) {
+          log.warn(
+              "Received unexpected heartbeat payload, perhaps the message subject {} is accidentally reused: {}",
+              HEARTBEAT_SUBJECT,
+              reply);
+        }
+      }
+
+      // Pass the message to the next handler, even though it's a heartbeat.
+      ctx.fireChannelRead(msg);
+    }
+
+    @Override
+    public void userEventTriggered(final ChannelHandlerContext ctx, final Object evt) {
+      if (!(evt instanceof final IdleStateEvent idleStateEvent)) {
+        return;
```

Comment:
:question: Should we not forward it to the next handler? :thinking:

Reply:
Potentially we could, but I don't think there's any other handler that consumes it at the moment. 
The cardinality of these events is not that much as they are triggered only when the channel is Idle.